### ISSUE TITLE
[RFC] Communicating failures in our APIs

### ISSUE THREAD

**Issue by MitjaBezensek** [tldraw/tldraw#3637]
## Problem 
Our API currently does not do a good job communicating failures to the consumers. This can lead to using the API in an overly optimistic way, thinking that the API call succeeded. 

The only way to prevent these kinds of issues at the moment is to do guard checks before / after calling the API, which is problematic since the are not really discoverable. You need to dig into the source code to see the ways in which API calls can fail. 

Another problem is that the before checks are also not enough due to the multiplayer nature of the library. The state can change between the time you do the check and the time you perform the action.

For an example see the issue we currently have with [creating shapes when max shapes](https://github.com/tldraw/tldraw/pull/3458) has already been reached. The consumers of `createShapes` have no way of knowing that this problem occurred, since the `createShapes` method returns the editor instance in any case. This caused issues for us, even though we are a lot more familiar with our APIs, but it is probably  worse for somebody that is not as familiar with the library.

### Do nothing

We can decide to avoid addressing this problem and just continue swallowing most of the errors. Interestingly we do throw errors in a couple of places.

```typescript
createShapes<T extends TLUnknownShape>(
shapes: OptionalKeys<TLShapePartial<T>, 'id'>[]
): this
```

### Throwing errors

API would throw errors as they are encountered. We could define different error types so that consumers get some additional information about what went wrong.

```typescript
	/**
 * 
 * ....
 *
 * @throws {@link MaxShapesReachedError}
 * Thrown if the maximum number of shapes per page was reached.
 * @throws {@link SomeOtherError}
 * Thrown if the bad stuff happened.
 *
 * @public
 */
createShapes<T extends TLUnknownShape>(
shapes: OptionalKeys<TLShapePartial<T>, 'id'>[]
): this 
```

Pros:

- This would preserve chaining.
- People are familiar with error handling.

Cons:

- Using the API would mean having to try catch all the calls that can throw if you really want to use it the way it’s intended.
- TypeScript doesn’t have support for declaring the types of thrown errors, so the DX is not the best. Can maybe use [docs to specify what methods](https://tsdoc.org/pages/tags/throws/) throw. Not sure if we can enforce this though?
- Throwing errors for expected failures (like the max shapes) does not feel right.
- Still not fully discoverable - code editors would not complain if you don’t account for possible errors.
- Requires quite a bit of work for the end users.

### Returning errors

Similar to throwing errors, but we instead return them. This helps with the DX as we can more easily expose the types of errors that can occur in the return types of the methods.

```typescript
createShapes<T extends TLUnknownShape>(
shapes: OptionalKeys<TLShapePartial<T>, 'id'>[]
): this | MaxShapesReachedError | SomeOtherError | ...
```

Pro:

- Different failures are more easily discoverable.

Cons:

- We’d loose chaining. Could maybe bring it back with some syntax sugar (see [chain in Tiptap](https://tiptap.dev/docs/editor/api/editor#chain)).
- This approach does not seem to be used that much, so it might not be something people are used to.
- To me this feels like a worse version of Result object.

### Returning statuses

We could only return statuses like a `true` / `false`. Could also be a list ids of the created shapes or null / undefined if something went wrong,…

```typescript
createShapes<T extends TLUnknownShape>(
shapes: OptionalKeys<TLShapePartial<T>, 'id'>[]
): boolean
```

Pros:

- Not as big of a change as others and also something people are used to from other APIs.
- Not as discoverable / enforceable as some other options.

Cons:

- We’d loose chaining. Could maybe bring it back with some syntax sugar (see [chain in Tiptap](https://tiptap.dev/docs/editor/api/editor#chain)).
- Consumers still don’t know what went wrong.

### Result object

Using a Result type (similar to what we already use [internally](https://github.com/tldraw/tldraw/blob/main/packages/utils/src/lib/control.ts) in some places). I did a quick exploration of this for `createShape` and `createShapes` method [here](https://github.com/tldraw/tldraw/pull/3622).

```typescript
createShapes<T extends TLUnknownShape>(
shapes: OptionalKeys<TLShapePartial<T>, 'id'>[]
): EditorResult<void, CreateShapeError>
```

Pros:

- I really like this pattern.
- Not as invasive as some other options since it would not change the flow of the program.

Cons:

- We’d loose chaining. Could maybe bring it back with some syntax sugar (see [chain in Tiptap](https://tiptap.dev/docs/editor/api/editor#chain)).
- It’s opinionated and might be too much to pass on to users.

**Comment by SomeHats** [comment]
a weird idea i've not put much thought into: `Editor.result`

#### How it would work

```ts
class Editor<ResultType = unknown> {
    // Editor has a `result` property. by default it's type is `unknown` and it's set to undefined
    result: ResultType = undefined

    // we have utilities that override this property, and return `this` with the type overridden
    protected returnOk<T>(value: T): Editor<OkResult<T>> {
        (this as any).result = Result.ok(value)
        return this as Editor<OkResult<T>>
    }
    protected returnErr<E>(error: E): Editor<ErrResult<E>> {
        (this as any).result = Result.err(error)
        return this as Editor<ErrResult<E>>
    }

    // Instead of just `return this`, operations that can fail use the helpers when returning.
    // the inferred return type here is `Editor<Result<Shape[], string>>`
    createShapes() {
        if (tooManyShapes) {
            return this.returnErr('too many shapes')
        }
        // ...
        return this.returnOk(createdShapes)
    }
}
```

#### Usage
```ts
// if i don't care about results:
editor.createShapes().updateShapes().whatever

// if i do care about results:
const result = editor.createShapes().result

if (result.ok) {
    console.log('created shapes:', result.value)
}
```

#### Pros
- We keep chaining
- Users can still choose to avoid the results (less impactful that it's opinionated)
- We can still return type-safe (ish) results

#### Cons
- Feels weird (?)
- Some type footguns possible:
```ts
const foo = editor.createShapes()
const bar = editor.updateShapes()

if (foo.result.ok) {
    // oh no, this is actually the `Result` from `updateShapes`
}
```

**Comment by MitjaBezensek** [comment]
> a weird idea i've not put much thought into: `Editor.result`

Yeah, it also crossed my mind 😄  It reminds me of working device statuses in low level programming.

**Comment by trygve-aaberge-adsk** [comment]
I like returning result objects. This makes it very clear that the operation can fail and how it can fail. It makes it easy to ensure you've handled all error cases. In contrast, throwing errors would require the user to look up what errors can be thrown in the documentation. And as you mentioned it's more type safe.

Result objects would also allow an API to be partially successful and return info about what was successful and what failed. E.g. when using `store.put` with multiple records and one of them is invalid, it could put the valid ones and return a result object describing which records were inserted and which failed.

As for the other options:
Do nothing: This is not viable IMO. That means you won't be able to handle any errors which would lead to inconsistent data and bugs.
Returning errors: This seems impractical compare to result objects to me. You'd have to do an instanceof Editor check to see if it's successful (or not instanceof Error). If you want any of the APIs to return data on success, it becomes a bit messy with a union of successful data and errors. It's just less clear when it's successful and when it fails.
Returning statuses: This makes it hard to handle the errors as you don't know what went wrong.

I think the main problem with result objects is that it makes it harder to read result data on a successful call because you're forced to handle the error. This may make it harder to play with the API before you necessarily care about handling all the errors. Though, for all the APIs where you don't need to read any result data (so all APIs that just return `Editor` today) it's not really a problem because you can just ignore the return value.

Another challenge is that uses might not be used to it as you said. It's a common pattern in some other functional languages though.

As for chaining I don't see it as a big issue. It's not that much of a problem to just repeat `editor.` before each API call IMO. And if you can add a function to do chaining like you linked to from Tiptap that would probably be fine. Additionally, supporting chaining by returning this for the API calls means you can't return any other data from any of the calls, which I think is a big limitation. I don't think it's worth sacrificing both better error handling and the possibility to return data just to get a bit easier chaining.

The suggestion from @SomeHats looks interesting, but I think it's too fragile that the result is stored in the editor and you potentially get the wrong result. It may not be obvious that the result is the same property in all returns, so I think it's too easy to make the mistake you posted an example of and get the wrong result. Getting the wrong result would be a pretty severe issue that shouldn't be this easy to run into.

So in conclusion I think the benefits of result object outweigh the drawbacks, and I don't think the alternative options proposed here are satisfactory.

You didn't post what the result would look like, but I'm guessing it's something like this?

```typescript
type EditorResult<Result, Error> = { ok: true, value: Result } | { ok: false, error: Error }
```

**Comment by MitjaBezensek** [comment]
> So in conclusion I think the benefits of result object outweigh the drawbacks, and I don't think the alternative options proposed here are satisfactory.

Thanks for your input on this. My main objection against result objects was that it might make the API a bit more complicated to use, but glad to hear that it doesn't feel too heavy to you

> 
> You didn't post what the result would look like, but I'm guessing it's something like this?
> 
> ```ts
> type EditorResult<Result, Error> = { ok: true, value: Result } | { ok: false, error: Error }
> ```

Yes, it would probably look something like the [result](https://github.com/tldraw/tldraw/blob/f93ff0d8b989170387896645da08794cd791b4cd/packages/utils/src/lib/control.ts#L4-L24) that we already use internally in some places.

> By the way, this link you have in the first post doesn't point to the correct line anymore since it points to the main branch and the file has changed since you posted the link. It looks like [this](https://github.com/tldraw/tldraw/blob/0d0d38361d9e0357af27bc7fa2c4a8573f382c72/packages/editor/src/lib/editor/Editor.ts#L6795) is the line it pointed to (I went back in the commit history to the date you posted the issue). To ensure links stay valid it's best to link to commits or tags, not branches (you can e.g. use the "Copy permalink" option in the drop down menu when you select a line).

Thanks for digging out the correct link, will try to remember to use the permalink option 👍
